starboard/android/shared/media_codec_bridge_eradicator.cc

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/android/shared/media_codec_bridge_eradicator.h"
+
+#include <pthread.h>
+
+#include "starboard/android/shared/jni_utils.h"
+#include "starboard/common/log.h"
+#include "starboard/common/once.h"
+
+namespace starboard {
+namespace android {
+namespace shared {
+
+SB_ONCE_INITIALIZE_FUNCTION(MediaCodecBridgeEradicator,
+                            MediaCodecBridgeEradicator::GetInstance);
+
+namespace {
+
+struct EradicateParam {
+  explicit EradicateParam(MediaCodecBridgeEradicator* eradicator,
+                          jobject j_media_codec_bridge,
+                          jobject j_reused_get_output_format_result)
+      : eradicator(eradicator),
+        j_media_codec_bridge(j_media_codec_bridge),
+        j_reused_get_output_format_result(j_reused_get_output_format_result) {}
+  MediaCodecBridgeEradicator* eradicator;
+  jobject j_media_codec_bridge;
+  jobject j_reused_get_output_format_result;
+};
+
+}  // namespace
+
+void MediaCodecBridgeEradicator::CheckEradicatingAndMaybeWait() {
+  ScopedLock scoped_lock(mutex_);
+  while (!j_media_codec_bridge_set_.empty()) {
+    condition_variable_.Wait();
+  }
+}
+
+bool MediaCodecBridgeEradicator::Eradicate(
```

Comment:
When the MediaCodec were stopped and destroyed asynchronously when a user chooses to stop a playback, do we notice any artifact?  For example, does the audio stay for a split second or the video flashes?

Reply:
I tested the code on Sabrina with a Samsung TV, there is no obvious difference compare to the production YouTube app.

Things I tested:
1. Start a video
2. Pause/Resume
3. bring the Kabuki control panel to find other videos, and then play the next videos
4. when video is playing, click the home button to put the app in the background